### PR TITLE
chore(aws-api-appsync): consolidate serde logic

### PR DESCRIPTION
**Pull request by jamesonwilliams**
Around the code base, Gson is used with various different
configurations. However, most uses require the same basic set of
serializers and deserializers, to handle custom Amplify and AppSync
types.

To address this, the serializers and deserializesr are consolidated into
the aws-api-appsync module. Construction of Gson instances is aligned to
use the common GsonFactory, now shared between the API and DataStore
categories.

The GsonFactory itself is augmented to provided different factory
methods, suitable for augmenting the set of adapters in a more
configurable way.

Resolves: https://github.com/aws-amplify/amplify-android/issues/848

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: aws-api-appsync/src/main/java/com/amplifyframework/api/aws/GsonFactory.java [hunk 1]
```diff
@@ -26,28 +26,62 @@
...
+     *          adapters
+     */
+    public static Gson create(@NonNull Map<Class<?>, Object> additionalAdapters) {
+        GsonBuilder builder = new GsonBuilder();
+        withAdditionalAdapters(builder, additionalAdapters);
+        withDeserializers(builder);
+        return create(builder);
+    }
+
     private static void withDeserializers(GsonBuilder builder) {
```
Comment (richardmcclellan) [inline aws-api-appsync/src/main/java/com/amplifyframework/api/aws/GsonFactory.java:78]:
```suggestion
    private static void withAdapters(GsonBuilder builder) {
```

**Review by raphkim — APPROVED** [review]

**Review by richardmcclellan — APPROVED** [review]
This is great!  Nice cleanup!

**Comment by richardmcclellan** [comment]
Now that `GsonFactory` is public, can we resolve [this feature request](https://github.com/aws-amplify/amplify-android/issues/815), which asks for an easy way to (de)serialize generated models?

**Comment by jamesonwilliams** [comment]
> Now that `GsonFactory` is public, will this also resolve [this feature request](https://github.com/aws-amplify/amplify-android/issues/815), which asks for an easy way to (de)serialize generated models?

It's a step in the right direction for https://github.com/aws-amplify/amplify-android/issues/815. But as of this PR, we still have hardcorded calls to create and use `GsonFactory.create()`, etc.

To complete #815, we'll need to abstract all of that and expose a single place to set the used Gson instance (perhaps, as a constructor/builder argument to the plugin(s)).

File: aws-datastore/src/main/java/com/amplifyframework/datastore/DataStoreGsonFactory.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import com.amplifyframework.datastore.appsync.ModelWithMetadataDeserializer;
+import com.amplifyframework.datastore.syncengine.TimeBasedUuidTypeAdapter;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+/**
+ * A utility to create Gson instances for the DataStore plugin.
+ */
+public final class DataStoreGsonFactory {
```
Comment (richardmcclellan) [inline aws-datastore/src/main/java/com/amplifyframework/datastore/DataStoreGsonFactory.java:31]: I'm a little confused by having two separate factories, `ApiGsonFactory`, and `DataStoreGsonFactory`.  From what I can tell, this is how you are using them: 
 - `ApiGsonFactory`
   - API request serialization
   - API response deserialization
   - DataStore response deserialization - this doesn't feel right!
 - `DataStoreApiFactory`
   - DataStore request serialization
   - SQLite (de)serialization of custom types


The DataStore `AppSyncClient` is using the API category under the hood, and hence is using the `ApiGsonFactory` for  all of the network operations - sync, mutate, subscribe.  


I would suggest either:
1. Just have one GsonFactory with everything
2. Have an `ApiGsonFactory` which is used for all API related (de)serialization in both API & DataStore, and a `CustomTypeGsonFactory`, which is used for (de)serialization of custom types to SQLite

Comment (jamesonwilliams) [reply aws-datastore/src/main/java/com/amplifyframework/datastore/DataStoreGsonFactory.java:31]: I should be able to achieve (1). It is the last change I was toying with, prior to putting up the latest version. Let me go back and see that option to its natural conclusion.

Comment (richardmcclellan) [reply aws-datastore/src/main/java/com/amplifyframework/datastore/DataStoreGsonFactory.java:31]: Awesome.  Other than that, the PR looks great.  Nice cleanup!

**Pull request merged**